src/compiler/checker.ts

```diff
@@ -280,6 +280,11 @@ namespace ts {
         const resolvingSignature = createSignature(undefined, undefined, undefined, emptyArray, anyType, /*typePredicate*/ undefined, 0, /*hasRestParameter*/ false, /*hasLiteralTypes*/ false);
         const silentNeverSignature = createSignature(undefined, undefined, undefined, emptyArray, silentNeverType, /*typePredicate*/ undefined, 0, /*hasRestParameter*/ false, /*hasLiteralTypes*/ false);
 
+        const unresolvedTypePredicate: void & { __unresolvedTypePredicate: void } = (() => {

```

Comment:
We generally don't have that issue with other cases like `noConstraintType`, especially if all access to the type predicate is gated through `getTypePredicateOfSignature`.

While I don't think it's strictly necessary, you could add another TypePredicateKind as a discriminant, but I think using an object whose shape (and hidden class) matches other valid values for that property will reduce polymorphism.